transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -84,16 +124,122 @@
     private volatile int handlerState = INIT;
 
     AbstractChannelHandlerContext(DefaultChannelPipeline pipeline, EventExecutor executor, String name,
-                                  boolean inbound, boolean outbound) {
+                                  Class<? extends ChannelHandler> handlerClass) {
         this.name = ObjectUtil.checkNotNull(name, "name");
         this.pipeline = pipeline;
         this.executor = executor;
-        this.inbound = inbound;
-        this.outbound = outbound;
+        this.executionMask = mask(handlerClass);
         // Its ordered if its driven by the EventLoop or the given Executor is an instanceof OrderedEventExecutor.
         ordered = executor == null || executor instanceof OrderedEventExecutor;
     }
 
+    /**
+     * Return the {@code executionMask}.
+     */
+    private static int mask(Class<? extends ChannelHandler> clazz) {
+        // Try to obtain the mask from the cache first. If this fails calculate it and put it in the cache for fast
+        // lookup in the future.
+        Map<Class<? extends ChannelHandler>, Integer> cache = MASKS.get();
+        Integer mask = cache.get(clazz);
+        if (mask == null) {
+            mask = mask0(clazz);
+            cache.put(clazz, mask);
+        }
+        return mask;
+    }
+
+    /**
+     * Calculate the {@code executionMask}.
+     */
+    private static int mask0(Class<? extends ChannelHandler> handlerType) {
+        int mask = MASK_EXCEPTION_CAUGHT;
```

Comment:
shouldn't this init as 0?

Reply:
@doom369 nope as `exceptionCaught` is part of `ChannelHandler`.